// @matter/model/dist/cjs/standard/elements/DishwasherMode.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DishwasherMode_exports = {};
__export(DishwasherMode_exports, {
  DishwasherMode: () => DishwasherMode
});
module.exports = __toCommonJS(DishwasherMode_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const DishwasherMode = (0, import_elements.ClusterElement)(
  {
    name: "DishwasherMode",
    id: 89,
    type: "ModeBase",
    classification: "application",
    pics: "DISHM",
    details: "This cluster is derived from the Mode Base cluster, defining additional mode tags and namespaced enumerated values for dishwasher devices.",
    xref: { document: "cluster", section: "8.3" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)({ name: "SupportedModes", id: 0, conformance: "M", xref: { document: "cluster", section: "8.3.5" } }),
  (0, import_elements.AttributeElement)({ name: "CurrentMode", id: 1, conformance: "M", xref: { document: "cluster", section: "8.3.5" } }),
  (0, import_elements.AttributeElement)({
    name: "StartUpMode",
    id: 2,
    conformance: "P",
    details: "If this attribute is supported, the device SHOULD initially set this to one of the supported modes that has the Normal tag associated with it. See the Mode Base cluster specification for full details about the StartUpMode attribute.",
    xref: { document: "cluster", section: "8.3.5.1" }
  }),
  (0, import_elements.AttributeElement)({ name: "OnMode", id: 3, conformance: "P", xref: { document: "cluster", section: "8.3.5" } }),
  (0, import_elements.DatatypeElement)({
    name: "ModeOptionStruct",
    type: "ModeOptionStruct",
    details: "The table below lists the changes relative to the Mode Base cluster for the fields of the ModeOptionStruct type. A blank field indicates no change.\nAt least one entry in the SupportedModes attribute shall include the Normal mode tag in the ModeTags field list.",
    xref: { document: "cluster", section: "8.3.4.1" }
  }),
  (0, import_elements.DatatypeElement)(
    { name: "ModeTag", type: "enum16" },
    (0, import_elements.FieldElement)({
      name: "Normal",
      id: 16384,
      details: "The normal regime of operation.",
      xref: { document: "cluster", section: "8.3.6.1.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "Heavy",
      id: 16385,
      details: "Mode optimized for washing heavily-soiled dishes.",
      xref: { document: "cluster", section: "8.3.6.1.2" }
    }),
    (0, import_elements.FieldElement)({
      name: "Light",
      id: 16386,
      details: "Mode optimized for light washing.",
      xref: { document: "cluster", section: "8.3.6.1.3" }
    })
  )
);
import_MatterDefinition.MatterDefinition.children.push(DishwasherMode);
