// @matter/model/dist/cjs/standard/elements/EventList.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EventList_exports = {};
__export(EventList_exports, {
  EventList: () => EventList
});
module.exports = __toCommonJS(EventList_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const EventList = (0, import_elements.AttributeElement)(
  {
    name: "EventList",
    id: 65530,
    type: "list",
    access: "R V",
    conformance: "P, M",
    isSeed: true,
    quality: "F",
    details: "Each instance of a cluster shall support this attribute. This attribute shall be a list of the event IDs of the events supported by the cluster instance.",
    xref: { document: "core", section: "7.13.6" }
  },
  (0, import_elements.FieldElement)({ name: "entry", type: "event-id" })
);
import_MatterDefinition.MatterDefinition.children.push(EventList);
