// @matter/model/dist/cjs/standard/elements/MicrowaveOvenMode.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MicrowaveOvenMode_exports = {};
__export(MicrowaveOvenMode_exports, {
  MicrowaveOvenMode: () => MicrowaveOvenMode
});
module.exports = __toCommonJS(MicrowaveOvenMode_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const MicrowaveOvenMode = (0, import_elements.ClusterElement)(
  {
    name: "MicrowaveOvenMode",
    id: 94,
    type: "ModeBase",
    classification: "application",
    pics: "MWOM",
    details: "This cluster is derived from the Mode Base cluster, defining additional mode tags and namespaced enumerated values for Microwave Oven devices.",
    xref: { document: "cluster", section: "8.12" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)({ name: "SupportedModes", id: 0, xref: { document: "cluster", section: "8.12.4" } }),
  (0, import_elements.AttributeElement)({ name: "CurrentMode", id: 1, xref: { document: "cluster", section: "8.12.4" } }),
  (0, import_elements.AttributeElement)({ name: "StartUpMode", id: 2, conformance: "X", xref: { document: "cluster", section: "8.12.4" } }),
  (0, import_elements.AttributeElement)({ name: "OnMode", id: 3, conformance: "X", xref: { document: "cluster", section: "8.12.4" } }),
  (0, import_elements.CommandElement)({ name: "ChangeToMode", id: 0, conformance: "X", xref: { document: "cluster", section: "8.12.5" } }),
  (0, import_elements.CommandElement)({ name: "ChangeToModeResponse", id: 1, conformance: "X", xref: { document: "cluster", section: "8.12.5" } }),
  (0, import_elements.DatatypeElement)(
    { name: "ModeTag", type: "enum16" },
    (0, import_elements.FieldElement)({
      name: "Normal",
      id: 16384,
      description: "The normal mode of operation",
      xref: { document: "cluster", section: "8.12.6.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "Defrost",
      id: 16385,
      description: "A mode optimized for defrosting foods",
      xref: { document: "cluster", section: "8.12.6.1" }
    })
  )
);
import_MatterDefinition.MatterDefinition.children.push(MicrowaveOvenMode);
