// @matter/model/dist/cjs/standard/elements/OvenCavityOperationalState.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OvenCavityOperationalState_exports = {};
__export(OvenCavityOperationalState_exports, {
  OvenCavityOperationalState: () => OvenCavityOperationalState
});
module.exports = __toCommonJS(OvenCavityOperationalState_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const OvenCavityOperationalState = (0, import_elements.ClusterElement)(
  {
    name: "OvenCavityOperationalState",
    id: 72,
    type: "OperationalState",
    classification: "application",
    pics: "OVENOPSTATE",
    details: "This cluster provides an interface for monitoring the operational state of an oven.",
    xref: { document: "cluster", section: "8.10" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 })
);
import_MatterDefinition.MatterDefinition.children.push(OvenCavityOperationalState);
