// @matter/model/dist/cjs/standard/elements/RvcCleanMode.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RvcCleanMode_exports = {};
__export(RvcCleanMode_exports, {
  RvcCleanMode: () => RvcCleanMode
});
module.exports = __toCommonJS(RvcCleanMode_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const RvcCleanMode = (0, import_elements.ClusterElement)(
  {
    name: "RvcCleanMode",
    id: 85,
    type: "ModeBase",
    classification: "application",
    pics: "RVCCLEANM",
    details: "This cluster is derived from the Mode Base cluster to define specifics for Robotic Vacuum Cleaner devices. It also defines a namespace for the cleaning type for these devices.",
    xref: { document: "cluster", section: "7.3" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)({ name: "SupportedModes", id: 0, xref: { document: "cluster", section: "7.3.6" } }),
  (0, import_elements.AttributeElement)({ name: "CurrentMode", id: 1, xref: { document: "cluster", section: "7.3.6" } }),
  (0, import_elements.AttributeElement)({ name: "StartUpMode", id: 2, conformance: "X", xref: { document: "cluster", section: "7.3.6" } }),
  (0, import_elements.AttributeElement)({ name: "OnMode", id: 3, conformance: "D", xref: { document: "cluster", section: "7.3.6" } }),
  (0, import_elements.DatatypeElement)({
    name: "ModeOptionStruct",
    type: "ModeOptionStruct",
    details: "The table below lists the changes relative to the Mode Base cluster for the fields of the ModeOptionStruct type. A blank field indicates no change.\nAt least one entry in the SupportedModes attribute shall include the Vacuum and/or the Mop mode tag in the ModeTags field list.",
    xref: { document: "cluster", section: "7.3.5.1" }
  }),
  (0, import_elements.DatatypeElement)(
    { name: "ModeChangeStatus", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "CleaningInProgress", id: 64, xref: { document: "cluster", section: "7.3.7.1" } })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ModeTag", type: "enum16" },
    (0, import_elements.FieldElement)({ name: "DeepClean", id: 16384, xref: { document: "cluster", section: "7.3.7.2" } }),
    (0, import_elements.FieldElement)({
      name: "Vacuum",
      id: 16385,
      details: "The device\u2019s vacuuming feature is enabled in this mode.",
      xref: { document: "cluster", section: "7.3.7.2.2" }
    }),
    (0, import_elements.FieldElement)({
      name: "Mop",
      id: 16386,
      details: "The device\u2019s mopping feature is enabled in this mode.",
      xref: { document: "cluster", section: "7.3.7.2.3" }
    })
  )
);
import_MatterDefinition.MatterDefinition.children.push(RvcCleanMode);
