// @matter/model/dist/esm/standard/elements/BridgedNodeDT.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const BridgedNodeDt = DeviceType(
  {
    name: "BridgedNode",
    id: 19,
    category: "Utility",
    classification: "utility",
    details: 'This defines conformance for a Bridged Node root endpoint. This endpoint is akin to a "read me first" endpoint that describes itself and any other endpoints that make up the Bridged Node. A Bridged Node endpoint represents a device on a foreign network, but is not the root endpoint of the bridge itself.',
    xref: { document: "device", section: "2.5" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 19, revision: 2 }], element: "attribute" })
  ),
  Requirement({
    name: "BridgedDeviceBasicInformation",
    id: 57,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "2.5.5" }
  }),
  Requirement({
    name: "PowerSourceConfiguration",
    id: 46,
    conformance: "BridgedPowerSourceInfo, D",
    element: "serverCluster",
    xref: { document: "device", section: "2.5.5" }
  }),
  Requirement({
    name: "PowerSource",
    id: 47,
    conformance: "BridgedPowerSourceInfo",
    element: "serverCluster",
    xref: { document: "device", section: "2.5.5" }
  })
);
MatterDefinition.children.push(BridgedNodeDt);
export {
  BridgedNodeDt
};
