// @matter/model/dist/esm/standard/elements/MicrowaveOvenMode.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  CommandElement as Command,
  DatatypeElement as Datatype,
  FieldElement as Field
} from "../../elements/index.js";
const MicrowaveOvenMode = Cluster(
  {
    name: "MicrowaveOvenMode",
    id: 94,
    type: "ModeBase",
    classification: "application",
    pics: "MWOM",
    details: "This cluster is derived from the Mode Base cluster, defining additional mode tags and namespaced enumerated values for Microwave Oven devices.",
    xref: { document: "cluster", section: "8.12" }
  },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute({ name: "SupportedModes", id: 0, xref: { document: "cluster", section: "8.12.4" } }),
  Attribute({ name: "CurrentMode", id: 1, xref: { document: "cluster", section: "8.12.4" } }),
  Attribute({ name: "StartUpMode", id: 2, conformance: "X", xref: { document: "cluster", section: "8.12.4" } }),
  Attribute({ name: "OnMode", id: 3, conformance: "X", xref: { document: "cluster", section: "8.12.4" } }),
  Command({ name: "ChangeToMode", id: 0, conformance: "X", xref: { document: "cluster", section: "8.12.5" } }),
  Command({ name: "ChangeToModeResponse", id: 1, conformance: "X", xref: { document: "cluster", section: "8.12.5" } }),
  Datatype(
    { name: "ModeTag", type: "enum16" },
    Field({
      name: "Normal",
      id: 16384,
      description: "The normal mode of operation",
      xref: { document: "cluster", section: "8.12.6.1" }
    }),
    Field({
      name: "Defrost",
      id: 16385,
      description: "A mode optimized for defrosting foods",
      xref: { document: "cluster", section: "8.12.6.1" }
    })
  )
);
MatterDefinition.children.push(MicrowaveOvenMode);
export {
  MicrowaveOvenMode
};
