// @matter/model/dist/esm/standard/elements/RvcCleanMode.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  DatatypeElement as Datatype,
  FieldElement as Field
} from "../../elements/index.js";
const RvcCleanMode = Cluster(
  {
    name: "RvcCleanMode",
    id: 85,
    type: "ModeBase",
    classification: "application",
    pics: "RVCCLEANM",
    details: "This cluster is derived from the Mode Base cluster to define specifics for Robotic Vacuum Cleaner devices. It also defines a namespace for the cleaning type for these devices.",
    xref: { document: "cluster", section: "7.3" }
  },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  Attribute({ name: "SupportedModes", id: 0, xref: { document: "cluster", section: "7.3.6" } }),
  Attribute({ name: "CurrentMode", id: 1, xref: { document: "cluster", section: "7.3.6" } }),
  Attribute({ name: "StartUpMode", id: 2, conformance: "X", xref: { document: "cluster", section: "7.3.6" } }),
  Attribute({ name: "OnMode", id: 3, conformance: "D", xref: { document: "cluster", section: "7.3.6" } }),
  Datatype({
    name: "ModeOptionStruct",
    type: "ModeOptionStruct",
    details: "The table below lists the changes relative to the Mode Base cluster for the fields of the ModeOptionStruct type. A blank field indicates no change.\nAt least one entry in the SupportedModes attribute shall include the Vacuum and/or the Mop mode tag in the ModeTags field list.",
    xref: { document: "cluster", section: "7.3.5.1" }
  }),
  Datatype(
    { name: "ModeChangeStatus", type: "enum8" },
    Field({ name: "CleaningInProgress", id: 64, xref: { document: "cluster", section: "7.3.7.1" } })
  ),
  Datatype(
    { name: "ModeTag", type: "enum16" },
    Field({ name: "DeepClean", id: 16384, xref: { document: "cluster", section: "7.3.7.2" } }),
    Field({
      name: "Vacuum",
      id: 16385,
      details: "The device\u2019s vacuuming feature is enabled in this mode.",
      xref: { document: "cluster", section: "7.3.7.2.2" }
    }),
    Field({
      name: "Mop",
      id: 16386,
      details: "The device\u2019s mopping feature is enabled in this mode.",
      xref: { document: "cluster", section: "7.3.7.2.3" }
    })
  )
);
MatterDefinition.children.push(RvcCleanMode);
export {
  RvcCleanMode
};
