// @matter/node/dist/esm/endpoints/device-energy-management.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  DeviceEnergyManagementServer as BaseDeviceEnergyManagementServer
} from "../behaviors/device-energy-management/DeviceEnergyManagementServer.js";
import {
  DeviceEnergyManagementModeServer as BaseDeviceEnergyManagementModeServer
} from "../behaviors/device-energy-management-mode/DeviceEnergyManagementModeServer.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { DeviceClassification } from "#model";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var DeviceEnergyManagementRequirements;
((DeviceEnergyManagementRequirements2) => {
  DeviceEnergyManagementRequirements2.DeviceEnergyManagementServer = BaseDeviceEnergyManagementServer;
  DeviceEnergyManagementRequirements2.DeviceEnergyManagementModeServer = BaseDeviceEnergyManagementModeServer;
  DeviceEnergyManagementRequirements2.server = {
    optional: {
      DeviceEnergyManagement: DeviceEnergyManagementRequirements2.DeviceEnergyManagementServer,
      DeviceEnergyManagementMode: DeviceEnergyManagementRequirements2.DeviceEnergyManagementModeServer
    },
    mandatory: {}
  };
})(DeviceEnergyManagementRequirements || (DeviceEnergyManagementRequirements = {}));
const DeviceEnergyManagementEndpointDefinition = MutableEndpoint({
  name: "DeviceEnergyManagement",
  deviceType: 1293,
  deviceRevision: 1,
  deviceClass: DeviceClassification.Utility,
  requirements: DeviceEnergyManagementRequirements,
  behaviors: SupportedBehaviors()
});
const DeviceEnergyManagementEndpoint = DeviceEnergyManagementEndpointDefinition;
export {
  DeviceEnergyManagementEndpoint,
  DeviceEnergyManagementEndpointDefinition,
  DeviceEnergyManagementRequirements
};
