// @project-chip/matter.js/dist/cjs/compat/behavior.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var behavior_exports = {};
__export(behavior_exports, {
  AccessControl: () => import_node.AccessControl,
  Behavior: () => import_node.Behavior,
  ClusterBehavior: () => import_node.ClusterBehavior,
  CommissioningBehavior: () => import_node.CommissioningServer,
  NetworkBehavior: () => import_node.NetworkBehavior,
  PartsBehavior: () => import_node.PartsBehavior,
  ProductDescriptionServer: () => import_node.ProductDescriptionServer,
  SessionsBehavior: () => import_node.SessionsBehavior
});
module.exports = __toCommonJS(behavior_exports);
var import_node = require("#node");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
