// @project-chip/matter.js/dist/cjs/compat/common.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var common_exports = {};
__export(common_exports, {
  FailsafeContext: () => import_protocol.FailsafeContext,
  FailsafeTimer: () => import_protocol.FailsafeTimer,
  ImplementationError: () => import_general.ImplementationError,
  InternalError: () => import_general.InternalError,
  Lifecycle: () => import_general.Lifecycle,
  MatterAggregateError: () => import_general.MatterAggregateError,
  MatterError: () => import_general.MatterError,
  MatterFlowError: () => import_general.MatterFlowError,
  NoProviderError: () => import_general.NoProviderError,
  NotImplementedError: () => import_general.NotImplementedError,
  ReadOnlyError: () => import_general.ReadOnlyError,
  StatusCode: () => import_types.StatusCode,
  TlvAttestation: () => import_protocol.TlvAttestation,
  UnexpectedDataError: () => import_general.UnexpectedDataError,
  ValidationDatatypeMismatchError: () => import_types.ValidationDatatypeMismatchError,
  ValidationError: () => import_types.ValidationError,
  ValidationMandatoryFieldMissingError: () => import_types.ValidationMandatoryFieldMissingError,
  ValidationOutOfBoundsError: () => import_types.ValidationOutOfBoundsError,
  validatorOf: () => import_types.validatorOf
});
module.exports = __toCommonJS(common_exports);
var import_general = require("#general");
var import_protocol = require("#protocol");
var import_types = require("#types");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
