// @project-chip/matter.js/dist/cjs/compat/interaction.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var interaction_exports = {};
__export(interaction_exports, {
  DEFAULT_MAX_PATHS_PER_INVOKE: () => import_types.DEFAULT_MAX_PATHS_PER_INVOKE,
  EventHandler: () => import_protocol.EventHandler,
  INTERACTION_MODEL_REVISION: () => INTERACTION_MODEL_REVISION,
  INTERACTION_PROTOCOL_ID: () => import_types.INTERACTION_PROTOCOL_ID,
  IncomingInteractionClientMessenger: () => import_protocol.IncomingInteractionClientMessenger,
  InteractionClient: () => import_protocol.InteractionClient,
  InteractionClientMessenger: () => import_protocol.InteractionClientMessenger,
  InteractionEndpointStructure: () => import_protocol.InteractionEndpointStructure,
  InteractionServer: () => import_protocol.InteractionServer,
  MessageType: () => import_protocol.MessageType,
  StatusResponseError: () => import_types.StatusResponseError,
  SubscriptionHandler: () => import_protocol.ServerSubscription,
  SubscriptionOptions: () => import_protocol.ServerSubscriptionConfig,
  attributePathToId: () => import_protocol.attributePathToId,
  canAttributePayloadBeChunked: () => import_protocol.canAttributePayloadBeChunked,
  chunkAttributePayload: () => import_protocol.chunkAttributePayload,
  clusterPathToId: () => import_protocol.clusterPathToId,
  commandPathToId: () => import_protocol.commandPathToId,
  compressAttributeDataReportTags: () => import_protocol.compressAttributeDataReportTags,
  decodeAttributeValueWithSchema: () => import_protocol.decodeAttributeValueWithSchema,
  decodeListAttributeValueWithSchema: () => import_protocol.decodeListAttributeValueWithSchema,
  decodeUnknownAttributeValue: () => import_protocol.decodeUnknownAttributeValue,
  decodeUnknownEventValue: () => import_protocol.decodeUnknownEventValue,
  encodeAttributePayload: () => import_protocol.encodeAttributePayload,
  encodeEventPayload: () => import_protocol.encodeEventPayload,
  eventPathToId: () => import_protocol.eventPathToId,
  expandPathsInAttributeData: () => import_protocol.expandPathsInAttributeData,
  genericElementPathToId: () => import_protocol.genericElementPathToId,
  normalizeAndDecodeAttributeData: () => import_protocol.normalizeAndDecodeAttributeData,
  normalizeAndDecodeEventData: () => import_protocol.normalizeAndDecodeEventData,
  normalizeAndDecodeReadAttributeReport: () => import_protocol.normalizeAndDecodeReadAttributeReport,
  normalizeAndDecodeReadEventReport: () => import_protocol.normalizeAndDecodeReadEventReport,
  normalizeAttributeData: () => import_protocol.normalizeAttributeData,
  normalizeEventData: () => import_protocol.normalizeEventData,
  sortAttributeDataByPath: () => import_protocol.sortAttributeDataByPath,
  structureReadAttributeDataToClusterObject: () => import_protocol.structureReadAttributeDataToClusterObject,
  validateReadAttributesPath: () => import_protocol.validateReadAttributesPath,
  validateReadEventPath: () => import_protocol.validateReadEventPath
});
module.exports = __toCommonJS(interaction_exports);
var import_model = require("#model");
var import_protocol = require("#protocol");
var import_types = require("#types");
__reExport(interaction_exports, require("@matter/types/protocol"), module.exports);
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const INTERACTION_MODEL_REVISION = import_model.Specification.INTERACTION_MODEL_REVISION;
