// @project-chip/matter.js/dist/cjs/compat/mdns.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var mdns_exports = {};
__export(mdns_exports, {
  MATTER_COMMISSIONER_SERVICE_QNAME: () => import_protocol.MATTER_COMMISSIONER_SERVICE_QNAME,
  MATTER_COMMISSION_SERVICE_QNAME: () => import_protocol.MATTER_COMMISSION_SERVICE_QNAME,
  MATTER_SERVICE_QNAME: () => import_protocol.MATTER_SERVICE_QNAME,
  MdnsBroadcaster: () => import_protocol.MdnsBroadcaster,
  MdnsInstanceBroadcaster: () => import_protocol.MdnsInstanceBroadcaster,
  MdnsScanner: () => import_protocol.MdnsScanner,
  MdnsServer: () => import_protocol.MdnsServer,
  SERVICE_DISCOVERY_QNAME: () => import_protocol.SERVICE_DISCOVERY_QNAME,
  getCommissioningModeQname: () => import_protocol.getCommissioningModeQname,
  getDeviceInstanceQname: () => import_protocol.getDeviceInstanceQname,
  getDeviceMatterQname: () => import_protocol.getDeviceMatterQname,
  getDeviceTypeQname: () => import_protocol.getDeviceTypeQname,
  getFabricQname: () => import_protocol.getFabricQname,
  getLongDiscriminatorQname: () => import_protocol.getLongDiscriminatorQname,
  getShortDiscriminatorQname: () => import_protocol.getShortDiscriminatorQname,
  getVendorQname: () => import_protocol.getVendorQname
});
module.exports = __toCommonJS(mdns_exports);
var import_protocol = require("#protocol");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
