// @project-chip/matter.js/dist/cjs/compat/net.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var net_exports = {};
__export(net_exports, {
  InterfaceType: () => import_general.InterfaceType,
  MAX_UDP_MESSAGE_SIZE: () => import_general.MAX_UDP_MESSAGE_SIZE,
  MockNetwork: () => import_general.MockNetwork,
  MockUdpChannel: () => import_general.MockUdpChannel,
  Network: () => import_general.Network,
  NetworkError: () => import_general.NetworkError,
  NetworkSimulator: () => import_general.NetworkSimulator,
  UdpInterface: () => import_general.UdpInterface,
  UdpMulticastServer: () => import_general.UdpMulticastServer
});
module.exports = __toCommonJS(net_exports);
var import_general = require("#general");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
