// @project-chip/matter.js/dist/cjs/compat/securechannel.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var securechannel_exports = {};
__export(securechannel_exports, {
  ChannelStatusResponseError: () => import_protocol.ChannelStatusResponseError,
  DEFAULT_NORMAL_PROCESSING_TIME_MS: () => import_protocol.DEFAULT_NORMAL_PROCESSING_TIME_MS,
  EXPECTED_CRYPTO_PROCESSING_TIME_MS: () => import_protocol.EXPECTED_CRYPTO_PROCESSING_TIME_MS,
  GeneralStatusCode: () => import_types.GeneralStatusCode,
  ProtocolStatusCode: () => import_types.ProtocolStatusCode,
  SecureChannelMessenger: () => import_protocol.SecureChannelMessenger,
  SecureChannelProtocol: () => import_protocol.SecureChannelProtocol,
  SecureChannelStatusMessageSchema: () => import_protocol.SecureChannelStatusMessageSchema,
  SecureMessageType: () => import_types.SecureMessageType,
  StatusReportOnlySecureChannelProtocol: () => import_protocol.StatusReportOnlySecureChannelProtocol
});
module.exports = __toCommonJS(securechannel_exports);
var import_protocol = require("#protocol");
var import_types = require("#types");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
