// @project-chip/matter.js/dist/cjs/compat/util.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var util_exports = {};
__export(util_exports, {
  AsyncObservable: () => import_general.AsyncObservable,
  Bytes: () => import_general.Bytes,
  Cache: () => import_general.Cache,
  EventEmitter: () => import_general.EventEmitter,
  MaybePromise: () => import_general.MaybePromise,
  Observable: () => import_general.Observable,
  Queue: () => import_general.DataReadQueue,
  camelize: () => import_general.camelize,
  createPromise: () => import_general.createPromise,
  deepCopy: () => import_general.deepCopy,
  isDeepEqual: () => import_general.isDeepEqual,
  isIPv6: () => import_general.isIPv6,
  onSameNetwork: () => import_general.onSameNetwork,
  serialize: () => import_general.serialize,
  singleton: () => import_general.singleton
});
module.exports = __toCommonJS(util_exports);
var import_general = require("#general");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
