// @project-chip/matter.js/dist/esm/LegacyControllerStore.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class LegacyControllerStore {
  #rootStorage;
  #sessionStorage;
  #caStorage;
  // Root certificate and Fabric
  #nodesStorage;
  // Holds list of nodes in root level and then sub levels with data per client node?
  /**
   * Create a new store.
   */
  constructor(storage) {
    this.#rootStorage = storage;
  }
  async erase() {
    await this.#sessionStorage?.clearAll();
    await this.#caStorage?.clearAll();
    await this.#nodesStorage?.clearAll();
  }
  async close() {
  }
  get sessionStorage() {
    if (!this.#sessionStorage) {
      this.#sessionStorage = this.#rootStorage.createContext("SessionManager");
    }
    return this.#sessionStorage;
  }
  get caStorage() {
    if (!this.#caStorage) {
      this.#caStorage = this.#rootStorage.createContext("RootCertificateManager");
    }
    return this.#caStorage;
  }
  get nodesStorage() {
    if (this.#nodesStorage === void 0) {
      this.#nodesStorage = this.#rootStorage.createContext("MatterController");
    }
    return this.#nodesStorage;
  }
  get fabricStorage() {
    return this.nodesStorage;
  }
  clientNodeStore(nodeId) {
    return this.nodesStorage.createContext(`node-${nodeId}`);
  }
}
export {
  LegacyControllerStore
};
