// @matter/protocol/dist/esm/interaction/Subscription.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger } from "#general";
const logger = Logger.get("Subscription");
class Subscription {
  #session;
  #id;
  #isClosed;
  #isCanceledByPeer;
  #criteria;
  constructor(session, id, criteria) {
    this.#session = session;
    this.#id = id;
    this.#criteria = criteria;
    this.#session.subscriptions.add(this);
    logger.debug(`Added subscription ${this.#id} to ${this.#session.name}`);
  }
  get id() {
    return this.#id;
  }
  get criteria() {
    return this.#criteria;
  }
  get isClosed() {
    return !!this.#isClosed;
  }
  get isCanceledByPeer() {
    return !!this.#isCanceledByPeer;
  }
  get session() {
    return this.#session;
  }
  /**
   * Update session state.  This probably is meaniningless except in a server context.
   */
  async update() {
  }
  /**
   * This flag is set on closed sessions when the close was initiated by the peer.
   */
  set isCanceledByPeer(value) {
    this.#isCanceledByPeer = value;
  }
  set isClosed(value) {
    this.#isClosed = value;
  }
  /**
   * Close the session.
   *
   * @param _graceful in a server context this means flush pending updates.  Not sure if applies to client
   * subscriptions
   */
  async close(_graceful = true) {
    this.#isClosed = true;
    this.#session.subscriptions.delete(this);
    logger.debug(`Removed subscription ${this.id} from ${this.#session.name}`);
  }
}
export {
  Subscription
};
