// @matter/node/dist/esm/behavior/state/transaction/Status.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { TransactionFlowError } from "./Errors.js";
var Status = /* @__PURE__ */ ((Status2) => {
  Status2["ReadOnly"] = "read only";
  Status2["Shared"] = "shared";
  Status2["Waiting"] = "waiting";
  Status2["Exclusive"] = "exclusive";
  Status2["CommittingPhaseOne"] = "committing phase one";
  Status2["CommittingPhaseTwo"] = "committing phase two";
  Status2["RollingBack"] = "rolling back";
  Status2["Destroyed"] = "destroyed";
  return Status2;
})(Status || {});
((Status2) => {
  function assert(transaction, acceptable, target) {
    if (!acceptable.includes(transaction.status)) {
      throw new TransactionFlowError(
        `Cannot transition transaction from ${formatStatus(transaction.status)} to ${formatStatus(target)}`
      );
    }
  }
  Status2.assert = assert;
  function formatStatus(status) {
    return `<${status}>`;
  }
  Status2.formatStatus = formatStatus;
})(Status || (Status = {}));
export {
  Status
};
