// @matter/nodejs/dist/esm/storage/StorageBackendDisk.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { fromJson, StorageError, SyncStorage, toJson } from "#general";
import { LocalStorage } from "node-localstorage";
class StorageBackendDisk extends SyncStorage {
  localStorage;
  isInitialized = false;
  constructor(path, clear = false) {
    super();
    this.localStorage = new LocalStorage(path);
    if (clear) this.clear();
  }
  get initialized() {
    return this.isInitialized;
  }
  initialize() {
    this.isInitialized = true;
  }
  close() {
    this.isInitialized = false;
  }
  clear() {
    this.localStorage.clear();
  }
  getContextBaseKey(contexts, allowEmptyContext = false) {
    const contextKey = contexts.join(".");
    if (!contextKey.length && !allowEmptyContext || contextKey.includes("..") || contextKey.startsWith(".") || contextKey.endsWith("."))
      throw new StorageError("Context must not be an empty and not contain dots.");
    return contextKey;
  }
  buildStorageKey(contexts, key) {
    if (!key.length) {
      throw new StorageError("Key must not be an empty string.");
    }
    const contextKey = this.getContextBaseKey(contexts);
    return `${contextKey}.${key}`;
  }
  get(contexts, key) {
    const value = this.localStorage.getItem(this.buildStorageKey(contexts, key));
    if (value === null) return void 0;
    return fromJson(value);
  }
  set(contexts, keyOrValues, value) {
    if (typeof keyOrValues === "string") {
      this.localStorage.setItem(this.buildStorageKey(contexts, keyOrValues), toJson(value));
    } else {
      for (const [key, value2] of Object.entries(keyOrValues)) {
        this.localStorage.setItem(this.buildStorageKey(contexts, key), toJson(value2));
      }
    }
  }
  delete(contexts, key) {
    this.localStorage.removeItem(this.buildStorageKey(contexts, key));
  }
  /** Returns all keys of a storage context without keys of sub-contexts */
  keys(contexts) {
    const contextKey = this.getContextBaseKey(contexts);
    const keys = [];
    const contextKeyStart = `${contextKey}.`;
    const len = contextKeyStart.length;
    for (const key of Object.keys(this.localStorage)) {
      if (key.startsWith(contextKeyStart) && !key.includes(".", len)) {
        keys.push(key.substring(len));
      }
    }
    return keys;
  }
  values(contexts) {
    const values = {};
    for (const key of this.keys(contexts)) {
      values[key] = this.get(contexts, key);
    }
    return values;
  }
  contexts(contexts) {
    const contextKey = this.getContextBaseKey(contexts, true);
    const startContextKey = contextKey.length ? `${contextKey}.` : "";
    const len = startContextKey.length;
    const foundContexts = new Array();
    for (const key of Object.keys(this.localStorage)) {
      if (key.startsWith(startContextKey)) {
        const subKeys = key.substring(len).split(".");
        if (subKeys.length === 1) continue;
        const context = subKeys[0];
        if (!foundContexts.includes(context)) {
          foundContexts.push(context);
        }
      }
    }
    return foundContexts;
  }
  clearAll(contexts) {
    const contextKey = this.getContextBaseKey(contexts, true);
    const startContextKey = contextKey.length ? `${contextKey}.` : "";
    for (const key of Object.keys(this.localStorage)) {
      if (key.startsWith(startContextKey)) {
        this.localStorage.removeItem(key);
      }
    }
  }
}
export {
  StorageBackendDisk
};
