// @project-chip/matter.js/dist/cjs/device/CachedClientNodeStore.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CachedClientNodeStore_exports = {};
__export(CachedClientNodeStore_exports, {
  CachedClientNodeStore: () => CachedClientNodeStore
});
module.exports = __toCommonJS(CachedClientNodeStore_exports);
var import_general = require("#general");
var import_protocol = require("#protocol");
var import_types = require("#types");
var import_ClientEndpointStore = require("./ClientEndpointStore.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("CachedClientNodeStore");
const VERSION_KEY = "__version__";
const MAX_EVENT_NUMBER_KEY = "__maxEventNumber__";
class CachedClientNodeStore extends import_protocol.PeerDataStore {
  #storage;
  #construction;
  #endpointStores = /* @__PURE__ */ new Map();
  #maxEventNumber = (0, import_types.EventNumber)(0);
  get construction() {
    return this.#construction;
  }
  constructor(storage, load = true) {
    super();
    this.#storage = storage;
    this.#construction = (0, import_general.Construction)(this, async () => {
      if (!load) {
        return;
      }
      this.#maxEventNumber = await this.#storage.get(MAX_EVENT_NUMBER_KEY, (0, import_types.EventNumber)(0));
      const knownParts = await this.#storage.contexts();
      for (const partId of knownParts) {
        await this.#loadEndpointStore(partId);
      }
    });
  }
  async #loadEndpointStore(partId) {
    const endpointNumber = (0, import_types.EndpointNumber)(parseInt(partId));
    const endpointStore = new import_ClientEndpointStore.ClientEndpointStore(this.#storage.createContext(partId));
    this.#endpointStores.set(endpointNumber, endpointStore);
    await endpointStore.construction;
  }
  #storeForEndpoint(endpointId) {
    this.#construction.assert();
    const existingStore = this.#endpointStores.get(endpointId);
    if (existingStore !== void 0) {
      return existingStore;
    }
    const store = new import_ClientEndpointStore.ClientEndpointStore(this.#storage.createContext(endpointId.toString()), false);
    this.#endpointStores.set(endpointId, store);
    return store;
  }
  get maxEventNumber() {
    return this.#maxEventNumber;
  }
  updateLastEventNumber(eventNumber) {
    this.#maxEventNumber = eventNumber;
    return this.#storage.set(MAX_EVENT_NUMBER_KEY, eventNumber);
  }
  retrieveAttribute(endpointId, clusterId, attributeId) {
    const store = this.#storeForEndpoint(endpointId);
    const clusterValues = store.get[clusterId];
    if (clusterValues === void 0) {
      return void 0;
    }
    const version = clusterValues[VERSION_KEY];
    if (version === void 0 || typeof version !== "number") {
      return void 0;
    }
    const attributeIdStr = attributeId?.toString();
    const data = clusterValues[attributeIdStr];
    const { value, attributeName } = data ?? {};
    if (value === void 0) {
      return void 0;
    }
    return {
      path: {
        endpointId,
        clusterId,
        attributeId,
        attributeName
      },
      value,
      version
    };
  }
  retrieveAttributes(endpointId, clusterId) {
    const store = this.#storeForEndpoint(endpointId);
    const clusterValues = store.get[clusterId];
    if (clusterValues === void 0) {
      return [];
    }
    const version = clusterValues[VERSION_KEY];
    if (version === void 0 || typeof version !== "number") {
      return [];
    }
    return Object.keys(clusterValues).filter((key) => key !== VERSION_KEY).map((key) => {
      const { value, attributeName } = clusterValues[key];
      return {
        path: { endpointId, clusterId, attributeId: (0, import_types.AttributeId)(parseInt(key)), attributeName },
        value,
        version
      };
    });
  }
  async persistAttributes(attributes) {
    const endpointDataMap = /* @__PURE__ */ new Map();
    for (const {
      path: { endpointId, clusterId, attributeId, attributeName },
      value,
      version
    } of attributes) {
      if (value === void 0) {
        logger.warn(`Ignoring undefined value for attribute ${attributeId} on endpoint ${endpointId}`);
        continue;
      }
      let endpointData = endpointDataMap.get(endpointId);
      if (endpointData === void 0) {
        endpointData = {};
        endpointDataMap.set(endpointId, endpointData);
      }
      let clusterData = endpointData[clusterId];
      if (clusterData === void 0) {
        clusterData = {};
        endpointData[clusterId] = clusterData;
      }
      clusterData[attributeId.toString()] = { value, attributeName };
      clusterData[VERSION_KEY] = version;
    }
    for (const [endpointId, endpointData] of endpointDataMap) {
      const store = this.#storeForEndpoint(endpointId);
      await store.set(endpointData);
    }
  }
  getClusterDataVersion(endpointId, clusterId) {
    const store = this.#storeForEndpoint(endpointId);
    const clusterValues = store.get[clusterId];
    return clusterValues?.[VERSION_KEY];
  }
  getClusterDataVersions(filterEndpointId, filterClusterId) {
    const versions = [];
    const filterClusterIdStr = filterClusterId?.toString();
    for (const [endpointId, store] of this.#endpointStores) {
      if (filterEndpointId !== void 0 && endpointId !== filterEndpointId) {
        continue;
      }
      const endpointData = store.get;
      for (const clusterId in endpointData) {
        if (filterClusterIdStr !== void 0 && clusterId !== filterClusterIdStr) {
          continue;
        }
        const version = endpointData[clusterId][VERSION_KEY];
        if (typeof version === "number") {
          versions.push({ endpointId, clusterId: (0, import_types.ClusterId)(parseInt(clusterId)), dataVersion: version });
        }
      }
    }
    return versions;
  }
}
